staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go

```diff
@@ -0,0 +1,725 @@
+	obj, shutdown := cfgCtl.configQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	func(obj interface{}) {
+		defer cfgCtl.configQueue.Done(obj)
+		if !cfgCtl.syncOne() {
+			cfgCtl.configQueue.AddRateLimited(obj)
+		} else {
+			cfgCtl.configQueue.Forget(obj)
+		}
+	}(obj)
+
+	return true
+}
+
+// syncOne attempts to sync all the API Priority and Fairness config
+// objects.  It either succeeds and returns `true` or logs an error
+// and returns `false`.
+func (cfgCtl *configController) syncOne() bool {
+	all := labels.Everything()
+	newPLs, err := cfgCtl.plLister.List(all)
+	if err != nil {
+		klog.Errorf("Unable to list PriorityLevelConfiguration objects: %s", err.Error())
+		return false
+	}
+	newFSs, err := cfgCtl.fsLister.List(all)
+	if err != nil {
+		klog.Errorf("Unable to list FlowSchema objects: %s", err.Error())
+		return false
+	}
+	err = cfgCtl.digestConfigObjects(newPLs, newFSs)
+	if err == nil {
+		return true
+	}
+	klog.Error(err)
+	return false
+}
+
+// cfgMeal is the data involved in the process of digesting the API
+// objects that configure API Priority and Fairness.  All the config
+// objects are digested together, because this is the simplest way to
+// cope with the various dependencies between objects.  The process of
+// digestion is done in four passes over config objects --- three
+// passes over PriorityLevelConfigurations and one pass over the
+// FlowSchemas --- with the work dvided among the passes according to
+// those dependencies.
+type cfgMeal struct {
+	cfgCtl *configController
+
+	newPLStates map[string]*priorityLevelState
+
+	// The sum of the concurrency shares of the priority levels in the
+	// new configuration
+	shareSum float64
+
+	// These keep track of which mandatory priority level config
+	// objects have been digested
+	haveExemptPL, haveCatchAllPL bool
+
+	// Buffered FlowSchema status updates to do.  Do them when the
+	// lock is not held, to avoid a deadlock due to such a request
+	// provoking a call into this controller while the lock held
+	// waiting on that request to complete.
+	fsStatusUpdates []fsStatusUpdate
+}
+
+// A buffered set of status updates for a FlowSchema
+type fsStatusUpdate struct {
+	flowSchema *fctypesv1a1.FlowSchema
+	condition  fctypesv1a1.FlowSchemaCondition
+	oldValue   fctypesv1a1.FlowSchemaCondition
+}
+
+// digestConfigObjects is given all the API objects that configure
+// cfgCtl and writes its consequent new configState.
+func (cfgCtl *configController) digestConfigObjects(newPLs []*fctypesv1a1.PriorityLevelConfiguration, newFSs []*fctypesv1a1.FlowSchema) error {
+	fsStatusUpdates := cfgCtl.lockAndDigestConfigObjects(newPLs, newFSs)
+	var errs []error
+	for _, fsu := range fsStatusUpdates {
+		enc, err := json.Marshal(fsu.condition)
+		if err != nil {
+			// should never happen because these conditions are created here and well formed
+			panic(fmt.Sprintf("Failed to json.Marshall(%#+v): %s", fsu.condition, err.Error()))
```

Comment:
Can error be returned here instead of panicing ?

Reply:
If this line is ever reached, either some very critical assumptions in this code are invalid or some bits have been silently been flipped in memory, and either way a panic is probably right.